File: Explore/2020May/Week2/C/remove-k-digits.c
char * removeKdigits(char * num, int k){
    int n = strlen(num);

    while (k > 0 && *num) 
    {
        while(*num == '0') 
        {
            num++;
            n--;
        }
        
        int i = 0;
        while((i < n - 1 )&&(num[i] <= num[i + 1])) 
            i++;

        for(int j = i; j < n - 1; j++)
            num[j] = num[j + 1];
            
        num[n - 1] = 0;
        n--;
        k--;
    }
    
    while(*num == '0')
        num++;

    if(!(*num)) 
        num = "0";
    
    return num;
}

File: Explore/2020May/Week2/Python/remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        result = []
        
        for i in num:
            while k and result and result[-1] > i:
                result.pop()
                k -= 1
            
            result.append(i)
            
        while k:
            result.pop()
            k -= 1
        
        return ''.join(result).lstrip('0') or '0'
